### PR TITLE
groundwork: repair procs

### PR DESCRIPTION
**Pull request by Karmic-Skink**
## About The Pull Request
Contains procs for repairing the ship overmap status

## Why It's Good For The Game
Precursor for [REDACTED]
Also for admin busing I guess

## Changelog
:cl:
add: Added some groundwork procs for later features
admin: Procs available easier overmap ship repairs
/:cl:

### CODE REVIEW

File: nsv13/code/modules/overmap/armour/armour_quadrant.dm [hunk 1]
```diff
@@ -66,4 +66,85 @@
 		if(L["current_armour"] <= L["max_armour"]/10) //Very close to buckling
 			add_overlay(L["name"])
 
-#undef ARMOUR_DING
\ No newline at end of file
+#undef ARMOUR_DING
+
+//Repair Procs
+
+/obj/structure/overmap/proc/full_repair_instant() //Admin Override
```
Comment (Kmc2000) [inline nsv13/code/modules/overmap/armour/armour_quadrant.dm:73]:
```suggestion
/obj/structure/overmap/proc/repair_full() //Admin Override
```

File: nsv13/code/modules/overmap/armour/armour_quadrant.dm [hunk 2]
```diff
@@ -66,4 +66,85 @@
...
+		priority_announce("Ship structural integrity restored to acceptable levels. ","Automated announcement ([src])")
+		structure_crit = FALSE
+		structure_crit_no_return = FALSE //Even override this
+	if(use_armour_quadrants) //Full armour plates
+		armour_quadrants["forward_port"]["current_armour"] = armour_quadrants["forward_port"]["max_armour"]
+		armour_quadrants["forward_starboard"]["current_armour"] = armour_quadrants["forward_starboard"]["max_armour"]
+		armour_quadrants["aft_port"]["current_armour"] = armour_quadrants["aft_port"]["max_armour"]
+		armour_quadrants["aft_starboard"]["current_armour"] = armour_quadrants["aft_starboard"]["max_armour"]
+
+/obj/structure/overmap/proc/modular_structure_repair(var/input, var/failure = 0) //Input is the amount you want repaired per call of this proc, failure is the chance the repair will go wrong
```
Comment (Kmc2000) [inline nsv13/code/modules/overmap/armour/armour_quadrant.dm:86]:
```suggestion
/obj/structure/overmap/proc/repair_structure(var/input, var/failure = 0) //Input is the amount you want repaired per call of this proc, failure is the chance the repair will go wrong
```

File: nsv13/code/modules/overmap/armour/armour_quadrant.dm [hunk 3]
```diff
@@ -66,4 +66,85 @@
...
+	obj_integrity += max_integrity * percentile
+	if(obj_integrity > max_integrity)
+		obj_integrity = max_integrity
+	if(structure_crit) //Check for structural crit
+		if(obj_integrity >= max_integrity * 0.2) //End structural crit if high enough
+			stop_relay(channel=CHANNEL_SHIP_FX)
+			priority_announce("Ship structural integrity restored to acceptable levels. ","Automated announcement ([src])")
+			structure_crit = FALSE
+
+/obj/structure/overmap/proc/modular_armour_repair_quadrant(var/input, var/failure = 0, var/bias = 50, var/quadrant) //Input is the amount you want repaired per call of this proc, failure is the chance the repair will go wrong, bias is the favour for structure damage vs armour damage
```
Comment (Kmc2000) [inline nsv13/code/modules/overmap/armour/armour_quadrant.dm:105]:
```suggestion
/obj/structure/overmap/proc/repair_quadrant(var/input, var/failure = 0, var/bias = 50, var/quadrant) //Input is the amount you want repaired per call of this proc, failure is the chance the repair will go wrong, bias is the favour for structure damage vs armour damage
```

File: nsv13/code/modules/overmap/armour/armour_quadrant.dm [hunk 4]
```diff
@@ -66,4 +66,85 @@
...
+				var/misrepair_quadrant = pick("forward_port", "forward_starboard", "aft_port", "aft_starboard")
+				armour_quadrants[misrepair_quadrant]["current_armour"] -= armour_quadrants[misrepair_quadrant]["max_armour"] * percentile
+				if(armour_quadrants[misrepair_quadrant]["current_armour"] < 0)
+					armour_quadrants[misrepair_quadrant]["current_armour"] = 0
+
+		armour_quadrants[quadrant]["current_armour"] += armour_quadrants[quadrant]["max_armour"] * percentile
+		if(armour_quadrants[quadrant]["current_armour"] > armour_quadrants[quadrant]["max_armour"])
+			armour_quadrants[quadrant]["current_armour"] = armour_quadrants[quadrant]["max_armour"]
+
+/obj/structure/overmap/proc/modular_armour_repair_all(var/input, var/failure = 0, var/bias = 50) //Input is the amount you want repaired per call of this proc, failure is the chance the repair will go wrong, bias is the favour for structure damage vs armour damage
```
Comment (Kmc2000) [inline nsv13/code/modules/overmap/armour/armour_quadrant.dm:127]:
```suggestion
/obj/structure/overmap/proc/repair_all_quadrants(var/input, var/failure = 0, var/bias = 50) //Input is the amount you want repaired per call of this proc, failure is the chance the repair will go wrong, bias is the favour for structure damage vs armour damage
```

**Review by Kmc2000 — CHANGES_REQUESTED** [review]

**Comment by francinum** [comment]
Could the last two procs not be combined, and the all behavior be triggered by `quadrant = null`?

File: nsv13/code/modules/overmap/armour/armour_quadrant.dm [hunk 5]
```diff
@@ -66,4 +66,85 @@
 		if(L["current_armour"] <= L["max_armour"]/10) //Very close to buckling
 			add_overlay(L["name"])
 
-#undef ARMOUR_DING
\ No newline at end of file
+#undef ARMOUR_DING
+
+//Repair Procs
+
+/obj/structure/overmap/proc/full_repair() //Admin Override
```
Comment (francinum) [inline nsv13/code/modules/overmap/armour/armour_quadrant.dm:73]: Could you add these as VV dropdown entries?

File: nsv13/code/modules/overmap/armour/armour_quadrant.dm [hunk 6]
```diff
@@ -148,3 +148,15 @@
...
+#define VV_HK_FULL_REPAIR "FullRepair"
+
+/obj/structure/overmap/vv_get_dropdown()
+	. = ..()
+	VV_DROPDOWN_OPTION(VV_HK_FULL_REPAIR, "Full Repair")
+
+/obj/structure/overmap/vv_do_topic(list/href_list)
+	. = ..()
+	if(href_list[VV_HK_FULL_REPAIR])
+		if(!check_rights(NONE)) //Hmm?
```
Comment (francinum) [inline nsv13/code/modules/overmap/armour/armour_quadrant.dm:156]:
```suggestion
		if(!check_rights(R_DEBUG)) //Hmm?
```

**Review by Kmc2000 — APPROVED** [review]

**Pull request merged**